documentation/intermediate-ca.md

```diff
@@ -0,0 +1,191 @@
+
+        csr = OpenSSL::X509::Request.new
+        csr.version = 0
+        csr.subject = root_cert.subject
+        csr.public_key = root_cert.public_key
+
+        ef = OpenSSL::X509::ExtensionFactory.new
+
+        ext_req = OpenSSL::X509::Attribute.new("extReq",
+          OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence([
+            ef.create_extension("keyUsage", "cRLSign, keyCertSign", true),
+            ef.create_extension("basicConstraints", "CA:TRUE", true)])]))
+
+        csr.add_attribute(ext_req)
+        csr.sign(ca_key, OpenSSL::Digest::SHA256.new)
+
+        puts csr.to_s
+      end
+    end
+    Puppet::Application::GenCACSR.new.run
+
+Use this CSR to generate the intermediate CA certificate that Puppet will be using.
+
+### Construct the CA certificate chain
+
+Puppet Server needs to present the full certificate chain to clients in order for the client to authenticate the server. The certificate chain is constructed by concatenating the CA certificates together, starting with the new intermediate CA certificate and descending to the root CA certificate.
+
+    cat puppet-intermediate-certificate.pem \
+        org-intermediate-certificate.pem \
+         org-root-certificate.pem > ca-bundle.pem
+
+### Verify the Puppet master host certificate against the CA bundle
+
+Before installing the new certificate bundle, verify that the bundle can be used to verify the existing host certificate on the CA server. If this step fails then the CA certificate bundle is invalid and cannot be installed.
+
+    openssl verify -CAfile ca-bundle.pem $(puppet master --configprint hostcert)
+
+### Update the Puppet CA certificate bundle
+
+    cp ca-bundle.pem /etc/puppetlabs/puppet/ssl/ca/ca_crt.pem
+
+### Update the Puppet agent CA certificate bundle
+
+The Puppet agent caches the CA certificate and will not automatically fetch the new CA certificate bundle from the master. This caveat applies to the Puppet CA server itself as well. On the CA server, copy the CA certificate bundle to the local CA certificate path:
+
+    cp /etc/puppetlabs/puppet/ssl/ca/ca_crt.pem /etc/puppetlabs/puppet/ssl/certs/ca.pem
+
+If any Puppet agents have already been created, copy the CA certificate bundle to `/etc/puppetlabs/puppet/ssl/certs/ca.pem`.
+
+Note that due to [PUP-6697](https://tickets.puppetlabs.com/browse/PUP-6697) Puppet agents are unable to download CA certificate bundles and will only save the first certificate in the bundle. Since the first certificate in the bundle will be the Puppet intermediate CA certificate, if the Puppet agent downloads and saves this single CA certificate it will not be able to complete verification of the Puppet Server certificate and will fail to connect.
+
+There may be other settings in the `webserver.conf` file; the other settings should be left alone.
+
+### Restart Puppet Server
+
+Once the new CA certificate is installed, restart pe-puppetserver.
+
+## Troubleshooting
```

Comment:
This section is awesome! Thank you! Our users thank you.

Reply:
Yes this is awesome!

Could we add prompts that tell the user where things are being run. [root@master]# , [root@agent]# 

The new administrator or someone who doesn't know PKI, but is required by their business to support it will thank you.